lib/iterator.js

```diff
@@ -0,0 +1,317 @@
+        return value;
+      }
+    }
+  }
+
+  includes(element) {
+    for (const value of this) {
+      if (value === element || (isNaN(value) && isNaN(element))) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  reduce(reducer, initialValue) {
+    let result = initialValue;
+
+    if (result === undefined) {
+      const next = this.next();
+      if (next.done) {
+        throw new TypeError(
+          'Reduce of consumed iterator with no initial value'
+        );
+      }
+      result = next.value;
+    }
+
+    for (const value of this) {
+      result = reducer(result, value);
+    }
+    return result;
+  }
+
+  some(predicate, thisArg) {
+    for (const value of this) {
+      if (predicate.call(thisArg, value)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  collectTo(CollectionClass) {
+    return new CollectionClass(this);
+  }
+
+  collectWith(obj, collector) {
+    this.forEach(element => collector(element, obj));
+  }
+
+  toArray() {
+    return Array.from(this);
+  }
+
+  map(mapper, thisArg) {
+    // eslint-disable-next-line no-use-before-define
+    return new MapIterator(this, mapper, thisArg);
+  }
+
+  filter(predicate, thisArg) {
+    // eslint-disable-next-line no-use-before-define
+    return new FilterIterator(this, predicate, thisArg);
+  }
+
+  flat(depth = 1) {
+    // eslint-disable-next-line no-use-before-define
+    return new FlatIterator(this, depth);
+  }
+
+  flatMap(mapper, thisArg) {
+    // eslint-disable-next-line no-use-before-define
+    return new FlatMapIterator(this, mapper, thisArg);
+  }
+
+  zip(...iterators) {
+    // eslint-disable-next-line no-use-before-define
+    return new ZipIterator(this, iterators);
+  }
+
+  join(...iterators) {
+    // eslint-disable-next-line no-use-before-define
+    return new JoinIterator(this, iterators);
+  }
+
+  take(amount) {
+    // eslint-disable-next-line no-use-before-define
+    return new TakeIterator(this, amount);
+  }
+
+  skip(amount) {
+    for (let i = 0; i < amount; i++) {
+      this.next();
+    }
+    return this;
+  }
+}
+
+class MapIterator extends Iterator {
+  constructor(base, mapper, thisArg) {
+    super(base);
+    this.mapper = mapper;
+    this.thisArg = thisArg;
+  }
+
+  next() {
+    const { done, value } = this.base.next();
+    return {
+      done,
+      value: done ? undefined : this.mapper.call(this.thisArg, value),
+    };
+  }
+}
+
+class FilterIterator extends Iterator {
+  constructor(base, predicate, thisArg) {
+    super(base);
+    this.predicate = predicate;
+    this.thisArg = thisArg;
+  }
+
+  next() {
+    for (const value of this.base) {
+      if (this.predicate.call(this.thisArg, value)) {
+        return { done: false, value };
+      }
+    }
+    return { done: true, value: undefined };
+  }
+}
+
+class FlatIterator extends Iterator {
+  constructor(base, depth) {
+    super(base);
+    this.depth = 0;
+    this.stack = new Array(depth + 1);
```

Comment:
Oh well, that's how it was before, lol =).

Also, I'd not be so sure about `new Array(length)` (as simple microbenchmarks show that preallocation wins by a great margin) though it indeed has to be tested on a per-case basis.

Reply:
@nechaido I'd like for it to be benchmarked if possible before we make the decision to pick one or the other implementation since I am quite sceptical about the performance benefits of *not* using `new Array(length)` and access by indices (when we have the maximal array length known at the start) compared to creating an empty array and using `Array.prototype.push()` and `Array.prototype.pop()` to add/remove the elements.

Reply:
@lundibundi I wasn't aware it was the previous implementation, hence a lot of comments, duh.

> Also, I'd not be so sure about new Array(length) (as simple microbenchmarks show that preallocation wins by a great margin)

If that hasn't changed over the last year, it doesn't: <https://v8project.blogspot.com/2017/09/elements-kinds-in-v8.html>.

Reply:
Keep in mind that a benchmark must push and pop items and access the last one repeatedly to emulate this particular use case.

Reply:
@aqrln @belochub I propose to rename `this.depth` to `this.currentDepth` as it may indeed be counterintuitive and leave the rest of the code as it is. Whenever someone proposes significant benchmarks that state that `[]` is faster or even equal - I'd be glad refactor this class.
Please leave your response ASAP, I would like to finally get over this PR.

Reply:
@nechaido sounds good to me